### PR TITLE
Query person_distinct_id2 not person_distinct_id

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

See issue: #8355 

The user in question cannot filter by `person` property even though the person and its property are present.

This was because the query in question uses `person_distinct_id` table when it should use person_distinct_id2

There is an `GET_TEAM_PERSON_DISTINCT_IDS` SQL string and a `GET_TEAM_PERSON_DISTINCT_IDS_NEW_TABLE` SQL string. They are chosen between using `ee/clickhouse/queries/person_distinct_id_query.py::get_team_distinct_ids_query` which requires the team id

If migration `0003_fill_person_distinct_id2` is complete then the application uses the new query. Otherwise, it uses the old one.

However, not every use of the two SQL queries comes through that method. Some code uses the SQL directly, and doesn't always have the team id available to pass for substitution.

## How did you test this code?

So far by creating the PR and seeing if any tests fail :)

### CODE REVIEW

File: ee/clickhouse/sql/person.py [hunk 1]
```diff
@@ -273,6 +273,7 @@
 HAVING argMax(is_deleted, version) = 0
 """
 
+# TODO are these templated correctly now?
```
Comment (macobo) [inline ee/clickhouse/sql/person.py:276]: Unrelated code?

File: ee/clickhouse/queries/person_distinct_id_query.py [hunk 2]
```diff
@@ -21,6 +21,21 @@ def get_team_distinct_ids_query(team_id: int) -> str:
         return substitute_params(GET_TEAM_PERSON_DISTINCT_IDS, {"team_id": team_id})
 
 
+def get_team_distinct_ids_query_without_substitution() -> str:
```
Comment (macobo) [inline ee/clickhouse/queries/person_distinct_id_query.py:24]: Q: Why not make the team_id argument optional to the other function? We're doubling our surface area here for APIs we need to maintain.

File: ee/clickhouse/sql/person.py (hunk 1 above)
Comment (pauldambra) [reply ee/clickhouse/sql/person.py:276]: Yep, pointers for myself to understand where it might need checking. Still very draft :)

File: ee/clickhouse/queries/person_distinct_id_query.py (hunk 2 above)
Comment (pauldambra) [reply ee/clickhouse/queries/person_distinct_id_query.py:24]: Yep, absolutely. And at least as bad if not worse... the queries worked but only because magically and unrelatedly code elsewhere was substituting team_id in.

When I flipped to start passing team_id along, I realised it was called from a lot of places I'd not have time to properly check them all _and_ do support hero. 

@liyiy has kindly agreed to pick it up... which may well mean going for a completely different solution even :)

Comment (liyiy) [reply ee/clickhouse/queries/person_distinct_id_query.py:24]: i'm happy to pick up where this PR left off if you want me to though it would make more sense if you continued this since you have the most context and I'll just take over a couple of support hero questions xD

**Comment by pauldambra** [comment]
@liyiy @hazzadous 

Would be good to review or close this :)

I've very little context on it other than chasing down a bug. So don't mind if the right thing is to close this :)

**Comment by neilkakkar** [comment]
Sounds crucial to get this in, specially since we've stopped populating the old table (more or less) on cloud, thus we'll always be wrong as long as we're using the old table!

File: ee/clickhouse/models/property.py [hunk 3]
```diff
@@ -143,9 +146,9 @@ def parse_prop_clauses(
                 final.append(f" AND {filter_query}")
             else:
                 # :TODO: (performance) Avoid subqueries whenever possible, use joins instead
-                # :TODO: Use get_team_distinct_ids_query instead when possible instead of GET_TEAM_PERSON_DISTINCT_IDS
+                assert isinstance(team_id, int)
```
Comment (hazzadous) [inline ee/clickhouse/models/property.py:149]: Same here re. `Optional`

**Review by hazzadous — CHANGES_REQUESTED** [review]
Seems we should be able to make team_id non-None? But maybe not.Otherwise looks sane.

File: ee/clickhouse/models/property.py [hunk 4]
```diff
@@ -82,10 +84,11 @@ def parse_prop_clauses(
                 final.append(filter_query)
                 params.update(filter_params)
             else:
+                assert isinstance(team_id, int)
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply ee/clickhouse/models/property.py]: Yep, was possible... had to pass team_id around from a bunch of places but not too terrible

File: ee/clickhouse/models/test/test_property.py [hunk 5]
```diff
@@ -543,18 +548,25 @@ def test_parse_prop_clauses_defaults(snapshot):
         }
     )
 
-    assert parse_prop_grouped_clauses(filter.property_groups, allow_denormalized_props=False) == snapshot
+    assert (
+        parse_prop_grouped_clauses(property_group=filter.property_groups, allow_denormalized_props=False, team_id=1)
```
Comment (macobo) [inline ee/clickhouse/models/test/test_property.py:552]: Nit: use selt.team.pk as above.

File: ee/clickhouse/models/test/test_property.py [hunk 6]
```diff
@@ -543,18 +548,25 @@ def test_parse_prop_clauses_defaults(snapshot):
...
             person_properties_mode=PersonPropertiesMode.USING_PERSON_PROPERTIES_COLUMN,
             allow_denormalized_props=False,
+            team_id=1,
         )
         == snapshot
     )
     assert (
         parse_prop_grouped_clauses(
-            filter.property_groups, person_properties_mode=PersonPropertiesMode.EXCLUDE, allow_denormalized_props=False
+            team_id=1,
```
Comment (macobo) [inline ee/clickhouse/models/test/test_property.py:566]: self.team.pk

File: ee/clickhouse/models/test/test_property.py [hunk 7]
```diff
@@ -543,18 +548,25 @@ def test_parse_prop_clauses_defaults(snapshot):
...
+        parse_prop_grouped_clauses(property_group=filter.property_groups, allow_denormalized_props=False, team_id=1)
+        == snapshot
+    )
     assert (
         parse_prop_grouped_clauses(
-            filter.property_groups,
+            property_group=filter.property_groups,
             person_properties_mode=PersonPropertiesMode.USING_PERSON_PROPERTIES_COLUMN,
             allow_denormalized_props=False,
+            team_id=1,
```
Comment (macobo) [inline ee/clickhouse/models/test/test_property.py:560]: self.team.pk

File: ee/clickhouse/models/property.py [hunk 8]
```diff
@@ -65,6 +66,7 @@
 
 def parse_prop_grouped_clauses(
     property_group: PropertyGroup,
+    team_id: int,
```
Comment (macobo) [inline ee/clickhouse/models/property.py:69]: Put this argument first - you're passing it that way everywhere anyways.

File: ee/clickhouse/models/property.py [hunk 9]
```diff
@@ -130,6 +134,7 @@ def is_property_group(group: Union[Property, "PropertyGroup"]):
 
 def parse_prop_clauses(
     filters: List[Property],
+    team_id: int,
```
Comment (macobo) [inline ee/clickhouse/models/property.py:137]: Put this argument first - you're passing it that way everywhere anyways.

File: ee/clickhouse/models/action.py [hunk 10]
```diff
@@ -11,6 +11,7 @@
 
 def format_action_filter(
     action: Action,
+    team_id: int,
```
Comment (macobo) [inline ee/clickhouse/models/action.py:14]: Put this argument first - you're passing it that way everywhere anyways.

File: ee/clickhouse/models/action.py [hunk 11]
```diff
@@ -93,10 +95,14 @@ def filter_event(
     return conditions, params
 
 
-def format_entity_filter(entity: Entity, prepend: str = "action", filter_by_team=True) -> Tuple[str, Dict]:
+def format_entity_filter(
+    entity: Entity, team_id: int, prepend: str = "action", filter_by_team=True
```
Comment (macobo) [inline ee/clickhouse/models/action.py:99]: Put `team_id` argument first - you're passing it that way everywhere anyways.

File: ee/clickhouse/models/test/test_action.py [hunk 12]
```diff
@@ -23,12 +23,12 @@ def _create_event(**kwargs) -> Event:
 
 
 def query_action(action: Action) -> Optional[List]:
-    formatted_query, params = format_action_filter(action, "")
+    formatted_query, params = format_action_filter(team_id=action.team.id, action=action, prepend="")
```
Comment (macobo) [inline ee/clickhouse/models/test/test_action.py:26]: Never do `thing.another.id` in an ORM when you could do `thing.another_id` - otherwise you're introducing a subquery where one isn't needed.

Use `action.team_id` here.

File: ee/clickhouse/models/test/test_action.py [hunk 13]
```diff
@@ -23,12 +23,12 @@ def _create_event(**kwargs) -> Event:
...
 
     query = ACTION_QUERY.format(action_filter=formatted_query)
 
     if query:
-        return sync_execute(query, {"team_id": action.team_id, **params})
+        return sync_execute(query, {"team_id": action.team.id, **params})
```
Comment (macobo) [inline ee/clickhouse/models/test/test_action.py:31]: Use `action.team_id` here.

**Review by macobo — COMMENTED** [review]
Some minor issues. Ping me when fixed and let's get this merged.

**Review by macobo — APPROVED** [review]
